### PR TITLE
Subhash (SuperUROP) - Semi-Supervised Learning

### PR DESCRIPTION
**Pull request by nsn116**

### CODE REVIEW

File: src/networks/competetive_net.py [hunk 1]
```diff
@@ -238,4 +238,107 @@ def compute_loss_against(self, opponent, input):
...
+        Assumes input is provided as a list and that the last column is an
+        integer column representing the class of the corresponding input
+        """
+        # Compute CrossEntropyLoss using real images
+        batch_size = input.size(0)
+        tensor = torch.Tensor(batch_size)
+        tensor.fill_(self.num_classes)
+        tensor = tensor.long()
+        fake_labels = to_pytorch_variable(tensor)
+        real = to_pytorch_variable(torch.ones(batch_size))
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py]: could you call them real_unsupervised_labels and fake_unsupervised_labels?? To make it more clear

File: src/networks/competetive_net.py [hunk 2]
```diff
@@ -238,4 +238,107 @@ def compute_loss_against(self, opponent, input):
...
+        fake = to_pytorch_variable(torch.zeros(batch_size))
+
+        # Real Loss
+        network_output = self.net(input)
+        network_output = network_output.view(network_output.shape[0], -1)
+        label_prediction = self.label_pred_layer(network_output).view(-1, self.num_classes + 1)
+        label_prediction_loss = self.label_pred_loss(label_prediction, labels)
+        outputs = self.adv_layer(network_output).view(-1)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py]: Outputs is the unsupervised output. Could you change the name to make it more clear?
unsupervised_pred or something like that?

File: src/networks/competetive_net.py [hunk 3]
```diff
@@ -238,4 +238,107 @@ def compute_loss_against(self, opponent, input):
...
+        validity = self.loss_function(outputs, real)
+        real_acc = np.mean(outputs.data.cpu().numpy())
+        d_loss_real = (label_prediction_loss + validity) / 2
+
+        pred = label_prediction.data.cpu().numpy()
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py]: try to change the variable names to make this process more clear (for these four following lines)

For example: d_acc is really hard to understand ;-)

File: src/networks/competetive_net.py [hunk 4]
```diff
@@ -238,4 +238,107 @@ def compute_loss_against(self, opponent, input):
...
+        d_acc = np.mean(pred_labels == gt)
+
+        # Fake Loss
+        z = noise(batch_size, self.data_size)
+        fake_images = opponent.net(z)
+        network_output = self.net(fake_images)
+        network_output = network_output.view(network_output.shape[0], -1)
+        label_prediction = self.label_pred_layer(network_output).view(-1, self.num_classes + 1)
+        label_prediction_loss = self.label_pred_loss(label_prediction, fake_labels)
+        outputs = self.adv_layer(network_output).view(-1)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py]: the same than before, change outputs name

File: src/networks/competetive_net.py [hunk 5]
```diff
@@ -238,4 +238,107 @@ def compute_loss_against(self, opponent, input):
...
+
+    def compute_loss_against(self, opponent: SSDiscriminatorNet, input, labels=None):
+        batch_size = input.size(0)
+
+        real_labels = to_pytorch_variable(torch.ones(batch_size))
+
+        z = noise(batch_size, self.data_size)
+
+        fake_images = self.net(z)
+        outputs = opponent.adv_layer(opponent.net(fake_images)).view(-1)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py]: outputs again :-)

File: src/networks/network_factory.py [hunk 6]
```diff
@@ -289,3 +298,59 @@ def forward(self, inputs, hidden=None):
             outputs[:, step, :] = output
 
         return outputs
+
+
+class SSGANPerceptronFactory(NetworkFactory):
```
Comment (jamaltoutouh) [inline src/networks/network_factory.py:372]: SSGANFourLayerPerceptronFactory to keep the names as it has been done

File: src/training/ea/lipizzaner_gan_trainer.py [hunk 7]
```diff
@@ -140,19 +140,23 @@ def train(self, n_iterations, stop_event=None):
             new_populations = {}
 
             # Create random dataset to evaluate fitness in each iterations
-            fitness_samples = self.generate_random_fitness_samples(self.fitness_sample_size)
+            fitness_input, fitness_labels = self.generate_random_fitness_samples(self.fitness_sample_size)
```
Comment (jamaltoutouh) [inline src/training/ea/lipizzaner_gan_trainer.py:156]: What have you changed fitness_samples to fitness_input? 
I like more fitness_samples name, isn't it?

File: src/training/ea/lipizzaner_gan_trainer.py [hunk 8]
```diff
@@ -485,28 +533,37 @@ def generate_random_fitness_samples(self, fitness_sample_size):
...
         if fitness_sample_size < batch_size:
-            return sampled_data[:fitness_sample_size]
+            return sampled_input[:fitness_sample_size], sampled_labels[:fitness_sample_size]
         else:
             fitness_sample_size -= batch_size
             while fitness_sample_size >= batch_size:
                 # Keep concatenate a full batch of data
-                curr_data, self.fitness_iterator = get_next_batch(self.fitness_iterator, self.fitness_loaded)
-                sampled_data = torch.cat((sampled_data, curr_data), 0)
+                curr_input, curr_labels, self.fitness_iterator = get_next_batch(
```
Comment (jamaltoutouh) [inline src/training/ea/lipizzaner_gan_trainer.py:650]: Why do you use _input?

File: src/networks/standalone_sgan/fid_score.py [hunk 9]
```diff
@@ -45,7 +48,11 @@ def __init__(self, imgs_original, batch_size=64, dims=2048, n_samples=10000, cud
         self.n_samples = n_samples
         self.cuda = cuda
         self.verbose = verbose
-        self.dims = 10    # For MNIST the dimension of feature map is 10
+        # For MNIST
+        # self.dims = 10
+
+        # For CIFAR
+        self.dims = 2048
```
Comment (jamaltoutouh) [inline src/networks/standalone_sgan/fid_score.py:55]: Could you add a new input parameter to select the data set? 
Thus, if self.dataset=='MNIST' or self.dataset=='CIFAR' you change dims, the way of loading the model, etc.

File: src/lipizzaner_master.py [hunk 10]
```diff
@@ -49,6 +49,9 @@ def run(self):
             clients = self.cc.settings['general']['distribution']['client_nodes']
         accessible_clients = self._accessible_clients(clients)
 
+        # NOTE: Uncomment this line to run Loss Function Based Diversity
+        # self.cc.settings['general']['distribution']['num_clients'] = len(accessible_clients)
+
```
Comment (jamaltoutouh) [inline src/lipizzaner_master.py:54]: Should we uncomment these lines?

File: src/main.py [hunk 11]
```diff
@@ -189,5 +268,9 @@ def generate_samples(args, cc):
     elif args.task == 'generate':
         cc = initialize_settings(args)
         generate_samples(args, cc)
+
+    elif args.task == 'ensemble-generator':
+        cc = initialize_settings(args)
+        ensemble_optimization(args, cc)
```
Comment (jamaltoutouh) [inline src/main.py:274]: I think that we should add a new task called 'test-classifiers' to allow us to load the discriminators and evaluate their performance by using most voted.
It should be something similar to 'generate' but loading the classifiers (discriminators+classification layer) and testing in on a given data set and a number of samples.

File: src/networks/competetive_net.py [hunk 12]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        self.classification_layer = classification_layer.cuda() if is_cuda_enabled() else classification_layer
+        self.mnist_28x28_conv = mnist_28x28_conv
+
+        cc = ConfigurationContainer.instance()
+        self.in_mean = cc.settings['network'].get('in_mean', 0.0)
+        self.in_std = cc.settings['network'].get('in_std', 1e-10)
+        self.in_std_decay_rate = cc.settings['network'].get('in_std_decay_rate', 0.0)
+        self.in_std_min = cc.settings['network'].get('in_std_min', 1e-10)
+        self.in_fake_decay = cc.settings['network'].get('in_fake_decay', False)
+        self.label_rate = cc.settings['dataloader'].get('label_rate', 1)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:283]: I think that the names of the variables should change since it is a little messy to understand the process.

File: src/networks/competetive_net.py [hunk 13]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        tensor.fill_(self.num_classes)
+        tensor = tensor.long()
+        fake_labels = to_pytorch_variable(tensor)
+
+        label_mask = self._get_labeled_mask(batch_size)
+
+        # Positive Label Smoothing
+        real = torch.Tensor(batch_size)
+        real.fill_(0.9)
+        real = to_pytorch_variable(real)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:366]: All these add-ons to the regular training should be documented. Because it has been a really long journey that you have traveled and it would be a pity if do not include all the steps in the report/paper.

File: src/networks/competetive_net.py [hunk 14]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        return d_loss_supervised + d_loss_unsupervised, None, accuracy
+
+
+class SSGeneratorNet(GeneratorNet):
+
+    def __init__(self, loss_function, num_classes, net, data_size,
+                 optimize_bias=True, fm=False):
+        GeneratorNet.__init__(self, loss_function, net, data_size, optimize_bias=optimize_bias)
+        self.num_classes = num_classes
+        self.fm = fm
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:438]: caould you named use_frature_matching??

File: src/networks/competetive_net.py [hunk 15]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        input = input + input_perturbation
+
+        if self.mnist_28x28_conv:
+            input = input.view(-1, 1, 28, 28)
+
+        network_output = self.classification_layer(self.net(input))
+        network_output = network_output.view(batch_size, -1)
+
+        # Real Supervised Loss
+        supervised_loss_function = CrossEntropyLoss(reduction='none')
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:383]: Why have you hardcoded the loss function? I saw that you have already created the configuration_container for them.
If you want, we can discuss it later.

File: src/networks/competetive_net.py [hunk 16]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        _logger = logging.getLogger(__name__)
+        _logger.info(f"Num Usable Labels: {num_usable_labels}")
+        loss_for_usable_labels = torch.sum(supervised_loss * label_mask)
+        label_prediction_loss = loss_for_usable_labels / num_usable_labels
+
+        # Real Unsupervised Loss
+        softmax_layer = Softmax()
+        probabilities = softmax_layer(network_output)
+        real_probabilities = -probabilities[:, -1] + 1
+        bce_loss = BCELoss()
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:395]: The same than before with the loss funciton

File: src/networks/competetive_net.py [hunk 17]
```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
...
+        else:
+            std = self.in_std
+        fake_image_perturbation = to_pytorch_variable(
+            torch.empty(fake_images.shape).normal_(mean=self.in_mean, std=std)
+        )
+        fake_images = fake_images + fake_image_perturbation
+
+        network_output = self.classification_layer(self.net(fake_images))
+        network_output = network_output.view(batch_size, -1)
+        label_prediction_loss = self.loss_function(network_output, fake_labels)
```
Comment (jamaltoutouh) [inline src/networks/competetive_net.py:426]: here you use the loss_function. Nice!!!

File: README.md [hunk 18]
```diff
@@ -7,19 +7,33 @@ Lipizzaner is a framework to train generative adversarial networks with gradient
 - Tom Schmiedlechner, Ignavier Ng Zhi Yong, Abdullah Al-Dujaili, Erik Hemberg, Una-May O'Reilly, “Lipizzaner: A System That Scales Robust Generative Adversarial Network Training,” NeurIPS 2018 Workshop on System for Machine Learning, 2018.
   - The configuration files to run the experiments performed in this paper can be found in the following path  `./src/configuration/neurips2018/`
 
-### Dev environment setup
+### Setup
 
-##### Conda
+1. Install Docker
```
[the remark this answers is not in the document]
Comment (nsn116) [reply README.md:12]: Yep, that is correct Jamal

File: src/data/data_loader.py [hunk 19]
```diff
@@ -35,21 +35,24 @@ def __init__(self, dataset, use_batch=True, batch_size=100, n_batches=0, shuffle
...
+                               train=train,
                                transform=self.transform(),
                                download=True)
 
         if self.sampling_ratio >= 1:
             self.sampler = None
         else:
-            set_random_seed(self.cc.settings['general']['seed'] + self.cell_number,
+            # set_random_seed(self.cc.settings['general']['seed'] + self.cell_number,
+            #                 self.cc.settings['trainer']['params']['score']['cuda'])
```
[the remark this answers is not in the document]
Comment (nsn116) [reply src/data/data_loader.py:54]: Oh, at the time, I think I was receiving errors about being unable to find client_id in the settings, but this has been fixed since in the latest version of master. Now that I have merged the latest version of master into my branch I will uncomment this and use the latest version!

File: src/lipizzaner_master.py (hunk 10 above)
Comment (nsn116) [reply src/lipizzaner_master.py:54]: Definitely Jamal, now that we are seeing improvements with it!

File: src/networks/competetive_net.py (hunk 13 above)
Comment (nsn116) [reply src/networks/competetive_net.py:366]: Definitely Jamal. They are currently on our SuperUROP Diary Google Doc but I'll add comments here to explain why we needed it

File: src/configuration/quickstart/cifar.yml [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+  name: lipizzaner_gan
+  n_iterations: 200
+  calculate_net_weights_dist: True
+  # independent_probability, exact_proportion
+  mixture_generator_samples_mode: exact_proportion
+  params:
+    population_size: 1
+    tournament_size: 2 
+    n_replacements: 1
+    default_adam_learning_rate: 0.0003
```
[the remark this answers is not in the document]
Comment (nsn116) [reply src/configuration/quickstart/cifar.yml:11]: Oh, I left it because it was there originally in the quickstart config files. But now, I've copied the files into src/configuration/quickstart-semi-supervised-learning specifically for SSL and I've removed the default_adam_learning_rate there.

**Comment by nsn116** [comment]
Thank you Jamal, I have made all requested changes. I would just like to test these changes and make sure nothing has broken following which we can quickly review the changes and approve this PR into the public ALFA repo?
